lib/ecto/association.ex

```diff
@@ -175,6 +175,85 @@ defmodule Ecto.Association do
+    relation_list = Enum.reduce(through, {owner, [[schema: owner, out_key: nil]]}, fn assoc, {owner, table_list} ->
+      refl = owner.__schema__(:association, assoc)
+      [owner_entry | table_list] = table_list
+
+      table_list = case refl do
+        %{join_through: join_through, join_keys: join_keys, join_where: join_where, where: where} ->
+          [{owner_join_key, owner_key}, {related_join_key, related_key}] = join_keys
+
+          owner_entry = owner_entry ++ [in_key: owner_key]
+          join_entry = [schema: join_through, out_key: owner_join_key, in_key: related_join_key, where: join_where]
+          related_entry = [schema: refl.related, out_key: related_key, where: where]
+
+          [related_entry, join_entry, owner_entry | table_list]
+
+        _ ->
+          owner_entry = owner_entry ++ [in_key: refl.owner_key]
+          related_entry = [schema: refl.related, out_key: refl.related_key, where: refl.where]
+
+          [related_entry, owner_entry | table_list]
+      end
+
+      {refl.related, table_list}
+    end)
+    |> elem(1)
+
+    # Filter out the joins which we don't need to use
+    filtered_list = Enum.with_index(relation_list)
+    |> Enum.filter(fn
+      # We always kee p the first table in the chain since it's our source table for the query
+      {_, 0} -> true
+
+      {rel, _} ->
+        condition = Keyword.get(rel, :where, [])
+        in_key = Keyword.get(rel, :in_key)
+        out_key = Keyword.get(rel, :out_key)
+
+        # If the condition is not empty we need to join to the table. Otherwise if the in_key and out_key is the same
+        # then this join is redundant.
+        in_key != out_key || condition != []
+    end)
+    |> Enum.map(&elem(&1, 0))
+    # Slice out the last table in the list since it is the owner table. We don't actually need to join it back.
+    |> Enum.slice(0..-2)
+
+    [source | joins] = filtered_list
+
+    source_schema = Keyword.get(source, :schema)
+    query = from(s in source_schema)
+
+    # We need to create the query by joining all the tables, and also we need the out_key of the final table to use
+    # for the final WHERE clause with values.
+    {_, query, _, dest_out_key} = Enum.reduce(joins, {source, query, 0, nil}, fn curr_rel, {prev_rel, query, counter, _} ->
+      related_queryable = Keyword.get(curr_rel, :schema)
+      out_key = Keyword.get(prev_rel, :out_key)
+      in_key = Keyword.get(curr_rel, :in_key)
+      where = Keyword.get(curr_rel, :where)
+
+      dest_out_key = Keyword.get(curr_rel, :out_key)
+
+      next =  join(query, :inner, [{src, counter}], dest in ^related_queryable, on: field(src, ^out_key) == field(dest, ^in_key))
+        |> combine_joins_query(where, counter + 1)
+
+      {curr_rel, next, counter + 1, dest_out_key}
+    end)
+
+    final_bind = Ecto.Query.Builder.count_binds(query) - 1
+    values = List.wrap(values)
+    query
+    |> where([{dest, final_bind}], field(dest, ^dest_out_key) in ^values)
```

Comment:
check if values is a list or not and emit different queries. Some DBs cannot cache `in` queries.